metrics/src/metrics.rs

```diff
@@ -0,0 +1,272 @@
+/// operators can also monitor external errors (client_error) errors to help
+/// improve their APIs or other systems.
+#[derive(Debug)]
+pub enum RedRequestStatus {
+    Ok,
+    ClientError,
+    Error,
+}
+
+impl Display for RedRequestStatus {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Ok => write!(f, "ok"),
+            Self::ClientError => write!(f, "client_error"),
+            Self::Error => write!(f, "error"),
+        }
+    }
+}
+
+#[derive(Debug)]
+/// A REDMetric is a metric that tracks requests to some resource.
+///
+/// The [RED methodology](https://www.weave.works/blog/the-red-method-key-metrics-for-microservices-architecture/)
+/// stipulates you should track three key measures:
+///
+///  - Request Rate: (total number of requests / second);
+///  - Error Rate: (total number of failed requests / second);
+///  - Duration: (latency distributions for the various requests)
+///
+/// Using a `REDMetric` makes following this methodology easy because it handles
+/// updating the three components for you.
+pub struct RedMetric {
+    default_labels: Vec<KeyValue>,
+    requests: OTCounter<u64>,
+    duration: OTHistorgram<f64>,
+}
+
+impl RedMetric {
+    pub(crate) fn new(
+        requests: OTCounter<u64>,
+        duration: OTHistorgram<f64>,
+        mut default_labels: Vec<KeyValue>,
+    ) -> Self {
+        // TODO(edd): decide what to do if `labels` contains
+        // RED_REQUEST_STATUS_LABEL.
+        // RedMetric always has a status label.
+        default_labels.insert(
+            0,
+            KeyValue::new(RED_REQUEST_STATUS_LABEL, RedRequestStatus::Ok.to_string()),
+        );
+
+        Self {
+            requests,
+            duration,
+            default_labels,
+        }
+    }
+
+    /// Returns a new observation that will handle timing and recording an
+    /// observation the metric is tracking.
+    pub fn observation(
+        &'_ self,
+    ) -> RedObservation<impl Fn(RedRequestStatus, Duration, &[KeyValue]) + '_> {
+        // The recording call-back
+        let record = move |status: RedRequestStatus, duration: Duration, labels: &[KeyValue]| {
+            let labels = match labels.is_empty() {
+                // If there are no labels specified just borrow defaults
+                true => Cow::Borrowed(&self.default_labels),
+                false => {
+                    // Otherwise merge the provided labels and the defaults.
+                    // Note: provided labels need to go last so that they overwrite
+                    // any default labels.
+                    //
+                    // PERF(edd): this seems expensive to me.
+                    let mut new_labels: Vec<KeyValue> = self.default_labels.clone();
+                    new_labels.extend(labels.iter().cloned());
+                    Cow::Owned(new_labels)
+                }
```

Comment:
Matching on one bool is a bit clearer as an if/else, IMO :) And you can use `extend_from_slice` to make this a bit shorter, I don't think it makes it more performant tho. It might not be bad because it looks like `KeyValue` uses `Cow` under the hood... profiling needed! :)

```suggestion
            let labels = if labels.is_empty() {
                // If there are no labels specified just borrow defaults
                Cow::Borrowed(&self.default_labels)
            } else {
                // Otherwise merge the provided labels and the defaults.
                // Note: provided labels need to go last so that they overwrite
                // any default labels.
                //
                // PERF(edd): this seems expensive to me.
                let mut new_labels: Vec<KeyValue> = self.default_labels.clone();
                new_labels.extend_from_slice(labels);
                Cow::Owned(new_labels)
```